internal/servers/worker/internal/metric/proxy_ws.go

```diff
@@ -0,0 +1,123 @@
+package metric
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	proxyWebsocketSubsystem = "worker_proxy_websocket"
+)
+
+var (
+	activeConnCount = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "active_connections",
+			Help:      "Count of open websocket connections to Boundary workers.",
+		},
+	)
+
+	bytesReceived = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "received_bytes_total",
+			Help:      "Count of received bytes for Worker connections.",
+		},
+	)
+
+	bytesSent = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "sent_bytes_total",
+			Help:      "Count of sent bytes for Worker connections.",
+		},
+	)
+)
+
+func MustInitializeProxyMetrics(r prometheus.Registerer) {
+	if r == nil {
+		return
+	}
+	r.MustRegister(activeConnCount, bytesReceived, bytesSent)
+}
+
+func ProxyWsMetricsHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		nextRw := rw
+		if rwh, ok := rw.(http.Hijacker); ok {
+			nextRw = &metricsWrapper{ResponseWriter: rw, Hijacker: rwh}
+		}
+		next.ServeHTTP(nextRw, req)
+	})
+}
+
+type wsConn struct {
+	*sync.Once
+	net.Conn
+}
+
+func (wc *wsConn) Read(b []byte) (int, error) {
+	n, err := wc.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+	bytesReceived.Add(float64(len(b))) // The slice is always at least 4096 bytes long (`bufio.ReadWriter` default)
```

Comment:
This is a noteworthy point that should be discussed in this RFC.
The byte slice len will be at least 4096. It'll consist of the actual data and the rest (if not filled up by the data itself) padded by zeroes.

It's probably worth to discuss if we should remove that padding for our counter as we'll be over-reporting received bytes if we don't.

Reply:
🤦 I was so focused on this problem that I literally forgot the data was right there. Thanks for the great spot.

Reply:
Addressed in [6156a64](https://github.com/hashicorp/boundary/pull/1963/commits/6156a64961d22cce2c58ec9be30f5e06e0683237)